### ISSUE TITLE
[FEA]Optimize CAST from string to Integer  

### ISSUE THREAD

**Issue by chenrui17** [NVIDIA/spark-rapids#1432]
**Is your feature request related to a problem? Please describe.**
The current implementation of CAST from string to integer relies on executing numerous regex expressions to determine integer exist in a column and this is very expensive. 

**Describe the solution you'd like**
I check the cudf from [cudf convert_integers.cu](https://github.com/rapidsai/cudf/blob/branch-0.18/cpp/src/strings/convert/convert_integers.cu) ,and there is no implementation of function `is_integer` , so do i need file an issue in [cudf repo](https://github.com/rapidsai/cudf/issues/new/choose) for support this ?

**Describe alternatives you've considered**
None

**Additional context**
before discuss is about poor performance of string regular matches in cudf repo https://github.com/rapidsai/cudf/issues/7048

**Comment by revans2** [comment]
@chenrui17 sorry about responding to this so late, I have been out on vacation.  What extra information can you share with us about this particular query? From the sql you posted in the previous issue.

> my sql is : ```
sql("select cmatch,query,sum(pv) as real_pv from table where event_day=20201107 and cmatch in (225,222,223) and page_number=0 and (event_isremoraabnor = 0 or (event_isremoraabnor = 1 and adview_hasclick = 1)) group by cmatch,query").show```
and cmatch, pv,page_number is bigint type ,and other related column is string

it looks like `event_isremoraabnor` is being converted to an int twice and `adview_hasclick` is being converted once.  Is the actual text stored in these columns just "0" and "1" values, or are they likely to hold other values too? Do you know the average length of the strings involved?

How many rows are your working with? The GPU tends to scale sub-linearly so small amounts of data can appear to run rather slowly.

**Comment by chenrui17** [comment]
> @chenrui17 sorry about responding to this so late, I have been out on vacation. What extra information can you share with us about this particular query? From the sql you posted in the previous issue.
> 
> > my sql is : ` sql("select cmatch,query,sum(pv) as real_pv from table where event_day=20201107 and cmatch in (225,222,223) and page_number=0 and (event_isremoraabnor = 0 or (event_isremoraabnor = 1 and adview_hasclick = 1)) group by cmatch,query").show`
> > and cmatch, pv,page_number is bigint type ,and other related column is string
> 
> it looks like `event_isremoraabnor` is being converted to an int twice and `adview_hasclick` is being converted once. Is the actual text stored in these columns just "0" and "1" values, or are they likely to hold other values too? Do you know the average length of the strings involved?
> 
> How many rows are your working with? The GPU tends to scale sub-linearly so small amounts of data can appear to run rather slowly.

1. Now, the actual text stored in these columns just "0" and "1" values temporary, in future ,they are likely to hold other values.
2.rows are 4,317,440,921
![image](https://user-images.githubusercontent.com/33319780/103599321-d8935400-4f3f-11eb-9c4a-fb1c23cc27d9.png)

**Comment by chenrui17** [comment]
@revans2  I find that there are `isFloat` and `isInteger` API in cudf java , so  i attempt to replace `isFloat` and `isInteger` with matchesRe(regex) at GpuCast.scala , the details are as follow: 
![image](https://user-images.githubusercontent.com/33319780/103620577-b8c75480-4f6e-11eb-862f-f27a66980d43.png)
In my personal repo: https://github.com/chenrui17/spark-rapids/commit/2712e70fab9a6905e44b435eae095b21f705ced8

and then , There seems to be no improvement in performance, by using nsys , time consuming part matchesRe is missing, but the performance almost same,
call `matchesRe`:
![image](https://user-images.githubusercontent.com/33319780/103621404-19a35c80-4f70-11eb-9560-7a18af176cba.png)

call `isFloat or isInteger`:
![image](https://user-images.githubusercontent.com/33319780/103621497-45264700-4f70-11eb-8c20-9b145a1867e4.png)

**Comment by revans2** [comment]
This is promising, but not going to be 100% what we want/need.  The underlying implementation in Spark

https://github.com/apache/spark/blob/cc1d9d25fb4c2e4af912d6f9802de8f351c32deb/common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java#L1116-L1194

When `allowDecimal` is false matches a pattern very close to what `trim` and `isInteger` does

https://github.com/rapidsai/cudf/blob/6ebd264c5655de19150f4d93c97800424105dffe/cpp/include/cudf/strings/string.cuh#L44-L53

But `isFloat` will match a lot more cases that Spark does not support.

https://github.com/rapidsai/cudf/blob/6ebd264c5655de19150f4d93c97800424105dffe/cpp/include/cudf/strings/string.cuh#L74-L107

In addition Spark will check for overflows, which we do not fully support.  Because of this we would need to work with cudf to try and implement something similar.  I'll try to use a modified version of your patch to get an estimate of the impact it would have for cast operations so that we can then decide how to prioritize any ultimate fixes for this.

**Comment by revans2** [comment]
I ran some benchmarks with similar code/data that was used here.

I generated data using

```
spark.range(0, 4317440921L).selectExpr("IF(random() > 0.5, \"0\", \"1\") AS v").write.parquet("target/tmp")
```

I split the data into 30 files, each one about 18 MB (529 MB total about 20 GB uncompressed)

I then ran the test under a few different conditions (CPU, GPU code today, and GPU code modified to use `isInteger` and `isFloat`)

```diff
diff --git a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala
index 329ffd05..2e93d1ac 100644
--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala
@@ -329,21 +329,21 @@ case class GpuCast(
                 GpuColumnVector.getNonNestedRapidsType(dataType))
             case ByteType | ShortType | IntegerType | LongType =>
               // filter out values that are not valid longs or nulls
-              val regex = if (ansiMode) {
-                GpuCast.ANSI_CASTABLE_TO_INT_REGEX
+              val formatIsCorrect = if (ansiMode) {
+                trimmed.isInteger
               } else {
-                GpuCast.CASTABLE_TO_INT_REGEX
+                trimmed.isFloat
               }
-              val longStrings = withResource(trimmed.matchesRe(regex)) { regexMatches =>
+              val longStrings = withResource(formatIsCorrect) { formatIsCorrect =>
                 if (ansiMode) {
-                  withResource(regexMatches.all(DType.BOOL8)) { allRegexMatches =>
-                    if (!allRegexMatches.getBoolean) {
+                  withResource(formatIsCorrect.all(DType.BOOL8)) { allCorrect =>
+                    if (!allCorrect.getBoolean) {
                       throw new NumberFormatException(GpuCast.INVALID_INPUT_MESSAGE)
                     }
                   }
                 }
                 withResource(Scalar.fromNull(DType.STRING)) { nullString =>
-                  regexMatches.ifElse(trimmed, nullString)
+                  formatIsCorrect.ifElse(trimmed, nullString)
                 }
               }

```


Query
```
spark.time(spark.read.parquet("target/tmp").select(count(when(col("v") === 1, true))).show())
```

12 CPU cores and default configs.

CPU:
37.8 seconds

GPU Original:
8.5 seconds

GPU updated:
3.5 seconds

This shows that there is a lot of potential in improveing the performance of a string cast to an int and would also give us the oportunity to fix the overflow checking that we are lacking right now.

**Comment by revans2** [comment]
I filed https://github.com/rapidsai/cudf/issues/7080 for changes to cudf to be able to support this properly.  Depending on what level of compatibility we want we could just put in my patch above, all of the tests pass with it but there are some corner cases we would get wrong (which might be OK because the cast is off by default if we document these corner cases properly) I also found that we could shave off a few seconds from the GPU test if we change the regular expressions so there is no white space checks (they are already trimmed) and if we limit the number of `[0-9]` characters to 19, which is the maximum that a long can hold.

**Comment by revans2** [comment]
The cudf changes were merged in and https://github.com/NVIDIA/spark-rapids/pull/2049 is to take advantage of them, but a requirement for the original cudf code was missed.  I need to add in java APIS for `is_fixed_point`, but I don't think it will work in all cases, and I might need to make a few changes to is_integer instead to support what we want.

**Comment by jlowe** [comment]
Fixed by #6436.
